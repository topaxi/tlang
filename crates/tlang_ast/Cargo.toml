[package]
name = "tlang_ast"
version = "0.1.0"
edition.workspace = true

[dev-dependencies]
indoc = { workspace = true }
pretty_assertions = { workspace = true }
