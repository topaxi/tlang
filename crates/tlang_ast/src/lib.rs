pub mod node;
pub mod span;
pub mod symbols;
pub mod token;
