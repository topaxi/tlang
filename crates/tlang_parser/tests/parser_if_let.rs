use indoc::indoc;
use pretty_assertions::assert_eq;

use tlang_ast::{
    node::{self},
    symbols::SymbolId,
    token::Literal,
};

mod common;

#[test]
fn test_simple_if_let_statement() {
    let program = parse!(indoc! {"
        if let x = 1 {
            x
        }
    "});

    assert_eq!(
        program,
        node::new!(Program(vec![node::new!(ExpressionStatement(Box::new(
            node::new!(IfElse {
                condition: Box::new(node::new!(VariableDeclaration {
                    id: SymbolId::new(1),
                    pattern: Box::new(node::new!(Identifier("x".to_string()))),
                    expression: Box::new(node::new!(Literal(Literal::Integer(1)))),
                    type_annotation: None,
                })),
                then_branch: Box::new(node::new!(Block(
                    vec![],
                    Some(Box::new(node::new!(Identifier("x".to_string()))))
                ))),
                else_branch: None,
            })
        )))]))
    );
}

#[test]
fn test_if_let_statement_with_enum_matching() {
    let program = parse!(indoc! {"
        if let Option::Some(y) = x; {
            y
        }
    "});

    assert_eq!(
        program,
        node::new!(Program(vec![node::new!(ExpressionStatement(Box::new(
            node::new!(IfElse {
                condition: Box::new(node::new!(VariableDeclaration {
                    id: SymbolId::new(1),
                    pattern: Box::new(node::new!(EnumExtraction {
                        identifier: Box::new(node::new!(NestedIdentifier(vec![
                            "Option".to_string(),
                            "Some".to_string()
                        ]))),
                        elements: vec![node::new!(Identifier("y".to_string()))],
                        named_fields: false,
                    })),
                    expression: Box::new(node::new!(Identifier("x".to_string()))),
                    type_annotation: None,
                })),
                then_branch: Box::new(node::new!(Block(
                    vec![],
                    Some(Box::new(node::new!(Identifier("y".to_string()))))
                ))),
                else_branch: None,
            })
        )))]))
    );
}
