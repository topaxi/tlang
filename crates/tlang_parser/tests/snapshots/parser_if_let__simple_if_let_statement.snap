---
source: crates/tlang_parser/tests/parser_if_let.rs
expression: "parse!(indoc! { \"\n        if let x = 1 {\n            x\n        }\n    \" })"
---
Node(
  ast_node: Program([
    Node(
      ast_node: ExpressionStatement(Node(
        ast_node: IfElse(
          condition: Node(
            ast_node: VariableDeclaration(
              id: SymbolId(1),
              pattern: Node(
                ast_node: Identifier("x"),
                symbol_table: None,
                span: Span(
                  start: LineColumn(
                    line: 0,
                    column: 7,
                  ),
                  end: LineColumn(
                    line: 0,
                    column: 8,
                  ),
                ),
              ),
              expression: Node(
                ast_node: Literal(Integer(1)),
                symbol_table: None,
                span: Span(
                  start: LineColumn(
                    line: 0,
                    column: 11,
                  ),
                  end: LineColumn(
                    line: 0,
                    column: 12,
                  ),
                ),
              ),
              type_annotation: None,
            ),
            symbol_table: None,
            span: Span(
              start: LineColumn(
                line: 0,
                column: 0,
              ),
              end: LineColumn(
                line: 0,
                column: 0,
              ),
            ),
          ),
          then_branch: Node(
            ast_node: Block([], Some(Node(
              ast_node: Identifier("x"),
              symbol_table: None,
              span: Span(
                start: LineColumn(
                  line: 1,
                  column: 5,
                ),
                end: LineColumn(
                  line: 1,
                  column: 6,
                ),
              ),
            ))),
            symbol_table: None,
            span: Span(
              start: LineColumn(
                line: 0,
                column: 0,
              ),
              end: LineColumn(
                line: 0,
                column: 0,
              ),
            ),
          ),
          else_branch: None,
        ),
        symbol_table: None,
        span: Span(
          start: LineColumn(
            line: 0,
            column: 0,
          ),
          end: LineColumn(
            line: 2,
            column: 2,
          ),
        ),
      )),
      symbol_table: None,
      span: Span(
        start: LineColumn(
          line: 0,
          column: 0,
        ),
        end: LineColumn(
          line: 2,
          column: 2,
        ),
      ),
    ),
  ]),
  symbol_table: None,
  span: Span(
    start: LineColumn(
      line: 0,
      column: 0,
    ),
    end: LineColumn(
      line: 0,
      column: 0,
    ),
  ),
)
